### PR TITLE
PiAffg Tests

### LINKED ISSUE #108: Write tests for PiAffgProof
A `PiAffgProof` has six elements that must satisfy seven different constraints. Some ideas for tests:
- Swap each of the proof elements for a different, random element (of the correct type). Proof shouldn't verify.
- Try making the proof with inconsistent secret inputs (e.g. `x` is not used to make `D`). Proof either shouldn't build correctly or shouldn't verify.
- Try making the proof with consistent secret inputs that are out of range (e.g. `x` is actually used to make `D`, but it's larger than `2^ell`).
- Try verifying the proof with different setup or public parameters than were used to construct it.

Related issue: #44.

### PR DESCRIPTION
**Pull request by hridambasu**
Closes #108

### CODE REVIEW

File: src/zkp/piaffg.rs [hunk 1]
```diff
@@ -596,6 +603,353 @@ mod tests {
...
+            input.verifier_encryption_key.clone(),
+            input.prover_encryption_key.clone(),
+            input.original_ciphertext_verifier.clone(),
+            bad_transformed_ciphertext_verifier,
+            input.add_coeff_ciphertext_prover.clone(),
+            input.mult_coeff_exp,
+        );
+        assert!(proof.verify(&bad_input, &(), &mut transcript).is_err());
+
+        // Swap verifier setup parameters with a random [`Ciphertext`]
```
Comment (marsella) [inline src/zkp/piaffg.rs:717]: nit: fix comment -- this is the additive coefficient

File: src/zkp/piaffg.rs [hunk 2]
```diff
@@ -596,6 +603,353 @@ mod tests {
...
+        assert!(proof.verify(&bad_input, &(), &mut transcript).is_err());
+
+        Ok(())
+    }
+
+    #[test]
+    fn piaffg_proof_with_inconsistent_secret_inputs() -> Result<()> {
+        let mut rng = init_testing();
+        let x = random_plusminus_by_size(&mut rng, ELL);
+        let y = random_plusminus_by_size(&mut rng, ELL);
```
Comment (marsella) [inline src/zkp/piaffg.rs:756]: todo: Fix `y` range.

File: src/zkp/piaffg.rs [hunk 3]
```diff
@@ -596,6 +603,353 @@ mod tests {
...
+        let pk1 = decryption_key_1.encryption_key();
+
+        let X = CurvePoint::GENERATOR.multiply_by_scalar(&x)?;
+        let (Y, rho_y) = pk1
+            .encrypt(&mut rng, &y)
+            .map_err(|_| InternalError::InternalInvariantFailed)?;
+
+        let C = pk0.random_ciphertext(&mut rng);
+
+        let (D, rho) = {
```
Comment (marsella) [inline src/zkp/piaffg.rs:770]: suggestion: I would recommend using the Rust-canonical names from the `Input` type rather than these one-letter capitals.

File: src/zkp/piaffg.rs [hunk 4]
```diff
@@ -596,6 +603,353 @@ mod tests {
...
+                    random_paillier_affg_proof(&mut rng, &too_small, &too_small).unwrap();
+                assert!(bad_proof.verify(&input, &(), &mut transcript).is_err());
+                break;
+            }
+        }
+        Ok(())
+    }
+
+    #[test]
+    fn piaffg_proof_with_different_setup_parameters() -> Result<()> {
```
Comment (marsella) [inline src/zkp/piaffg.rs:631]: issue: please use the `transcript()` function throughout this test.

File: src/zkp/piaffg.rs [hunk 5]
```diff
@@ -596,6 +603,353 @@ mod tests {
...
+            let too_large = random_plusminus_by_size_with_minimum(
+                &mut rng,
+                ELL + EPSILON + 2,
+                ELL + EPSILON + 1,
+            )?;
+            let too_small = -too_large.clone();
+            let (bad_proof, input, mut transcript) =
+                random_paillier_affg_proof(&mut rng, &too_large, &too_large).unwrap();
+            if too_large.gt(&upper_bound) {
+                assert!(bad_proof.verify(&input, &(), &mut transcript).is_err());
```
Comment (marsella) [inline src/zkp/piaffg.rs:620]: issue: please use the `transcript()` function throughout this test.

File: src/zkp/piaffg.rs [hunk 6]
```diff
@@ -596,6 +603,353 @@ mod tests {
...
+        assert!(bad_proof.verify(&input, &(), &mut transcript()).is_err());
+
+        Ok(())
+    }
+
+    #[test]
+    fn negative_test_random_proof_elements() -> Result<()> {
+        let mut rng = init_testing();
+        let x = random_plusminus_by_size(&mut rng, ELL);
+        let y = random_plusminus_by_size(&mut rng, ELL);
```
Comment (marsella) [inline src/zkp/piaffg.rs:820]: todo: fix `y` range. This will technically always be valid because `ELL < ELL_PRIME` so you're generating an in-range value. But I think it's worth fixing just to avoid confusion in the future about what the expected range is. Plus, having the correct range will ensure that (over time) we randomly test elements across the whole thing including the edge cases.

File: src/zkp/piaffg.rs [hunk 7]
```diff
@@ -596,6 +603,353 @@ mod tests {
...
+        assert_ne!(secret.add_coeff_nonce_prover_key, random_nonce);
+        let bad_secret = PiAffgSecret::new(x, y, rho, random_nonce);
+        let bad_proof = PiAffgProof::prove(&input, &bad_secret, &(), &mut transcript(), &mut rng)?;
+        assert!(bad_proof.verify(&input, &(), &mut transcript()).is_err());
+
+        Ok(())
+    }
+
+    #[test]
+    fn negative_test_random_proof_elements() -> Result<()> {
```
Comment (marsella) [inline src/zkp/piaffg.rs:817]: issue: use the `transcript()` function throughout this proof.

**Review by marsella — CHANGES_REQUESTED** [review]
This looks good! My comments are mostly quite small -- the range test needs some work and `transcript()` needs to be used more consistently, but the enormous switching-out-elements tests all look great.

I would recommend you rebase this sooner rather than later--Omar changed the structure of these tests in #395 and it's different enough that I'd like to review this with those changes incorporated. You won't have to change the content of the tests but the structure will get wiggled around a bit; you can see the diff in the linked PR for details and feel free to ping @gatoWololo for more context.

File: src/zkp/piaffg.rs [hunk 8]
```diff
@@ -604,12 +608,437 @@ mod tests {
...
+            assert!(bad_proof.verify(input, &(), &mut transcript()).is_err());
+
+            Ok(())
+        };
+        with_random_paillier_affg_proof(rng2, &x, &y, f)?;
+        Ok(())
+    }
+
     #[test]
     fn test_paillier_affg_proof() -> Result<()> {
```
Comment (marsella) [inline src/zkp/piaffg.rs:1037]: I actually see this proof does some range testing, too. I prefer having a separate named proof that tests the range stuff like you have but just noting the duplication here.

**Review by marsella — CHANGES_REQUESTED** [review]

File: src/zkp/piaffg.rs [hunk 9]
```diff
@@ -604,12 +608,437 @@ mod tests {
...
+                let f: TestFn = |bad_proof, input| {
+                    assert!(bad_proof.verify(input, &(), &mut transcript()).is_err());
+                    Ok(())
+                };
+                with_random_paillier_affg_proof(&mut rng, &x_too_large, &y, f)?;
+                with_random_paillier_affg_proof(&mut rng, &x_too_small, &y, f)?;
+                break;
+            }
+        }
+        loop {
```
[the remark this answers is not in the document]
Comment (hridambasu) [reply src/zkp/piaffg.rs:635]: yes precisely. looping so that the too_large number is actually generated sometime. If I omit the loop, then there is no guarantee that the test will actually generate a too large or too small number that we want to test with. So then the sampling function will work but there is also the possibility that most of the times the number generated is less than the too large range. Then there is no point of the test.

File: src/zkp/piaffg.rs [hunk 10]
```diff
@@ -604,12 +608,437 @@ mod tests {
         with_random_paillier_affg_proof(rng, x, y, f)
     }
 
+    #[test]
+    fn piaffg_proof_with_consistent_secret_inputs_out_of_range() -> Result<()> {
+        let mut rng = init_testing();
+        let x_upper_bound = BigNumber::one() << (ELL + EPSILON);
+        let y_upper_bound = BigNumber::one() << (2 ^ (ELL_PRIME + EPSILON));
```
[the remark this answers is not in the document]
Comment (marsella) [reply src/zkp/piaffg.rs]: Oh, I see I was wrong here because I misunderstood how these were being used. I thought they were the sampling bounds for the valid inputs, but they're the minimum bounds for invalid inputs.

Please tell me when you think I'm wrong about something like this.

File: src/zkp/piaffg.rs [hunk 11]
```diff
@@ -604,12 +608,437 @@ mod tests {
...
+        };
+        with_random_paillier_affg_proof(rng2, &x, &y, f)?;
+        Ok(())
+    }
+
+    #[test]
+    fn piaffg_proof_with_inconsistent_secret_inputs() -> Result<()> {
+        let mut rng = init_testing();
+        let x = random_plusminus_by_size(&mut rng, ELL);
+        let y = random_plusminus_by_size(&mut rng, 2 ^ ELL_PRIME);
```
[the remark this answers is not in the document]
Comment (marsella) [reply src/zkp/piaffg.rs]: I'm still seeing incorrectly-exponented bounds throughout the tests. All the sampling functions with `by_size` in the name take the exponent as a parameter.

**Review by marsella — COMMENTED** [review]
Following up on some open comment threads. There are still some open threads from previous reviews that didn't get addressed. If you think they're not correct, please note why.

File: src/zkp/piaffg.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (hridambasu) [reply src/zkp/piaffg.rs:635]: Oh yeah you are right! Sorry about that, my bad.

**Review by marsella — APPROVED** [review]
Looks good, thank you!
